Fix Infinite Loading Screen + Fix Modsettings "int" functionality

EnderBoy9217 opened the pull request:
Previously if a server crashed while the client was in the map loading screen, all clients would be stuck in an infinite loading screen. (This was reproduced by multiple people)

This boots clients back to the menu if the game hasn't loaded within 30 seconds (server crash) and loads the main menu, it doesn't affect the campaign's "continue" feature either.

Video below (some footage sped up by 8x to reduce video size)

https://github.com/R2Northstar/NorthstarMods/assets/122132914/f1e1cc3f-bbd2-408e-9f0f-ef7ef89cf296


EDIT:
This adds a Northstar.Client section to modsettings
This also fixes an issue where modsettings cannot take an "int" ConVar
Closes: https://github.com/R2Northstar/Northstar/issues/489

uniboi commented:
Maybe open a dialog when returning to the main menu that tells the user that the loading has timed out

ASpoonPlaysGames commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,20 @@ void function UICodeCallback_ToggleInGameMenu()
+		UICodeCallback_LevelLoadingFinished( true )
+		Disconnect()
+	}
+	return
```
```suggestion
```
unnecessary `return` can be removed

ASpoonPlaysGames commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,20 @@ void function UICodeCallback_ToggleInGameMenu()
+	{
+		printt("CancelLevelLoadingInfinite disconnected because loading took too long")
+		UICodeCallback_LevelLoadingFinished( true )
+		Disconnect()
```
Would definitely be nice to show the user a dialog here that explains things (tbh the "Connection Timed Out" dialog in vanilla would work)

ASpoonPlaysGames commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,20 @@ void function UICodeCallback_ToggleInGameMenu()
 	}
 }
 
+void function CancelLevelLoadingInfinite()
```
Not a big fan of this function name tbh

EnderBoy9217 commented:
Added an error message for when this happens
![Desktop Screenshot 2023 06 17 - 11 37 35 46](https://github.com/R2Northstar/NorthstarMods/assets/122132914/912220e3-1873-4c95-ad95-0c60253290d7)

F1F7Y commented:
My main concern would be what if the loading actually takes more than 30 seconds?

Please also comment your code, scripts are already chaotic the way they are right now

uniboi commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,20 @@ void function UICodeCallback_ToggleInGameMenu()
 	}
 }
 
+void function CancelLevelLoading()
```
I'd say this is a better name
```suggestion
void function TimeoutHandler()
```

uniboi commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -204,7 +218,7 @@ bool function UICodeCallback_LevelLoadingStarted( string levelname )
 		return false
 #endif
-
+	thread CancelLevelLoading()
```
```suggestion
	thread TimeoutHandler()
```

uniboi commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,20 @@ void function UICodeCallback_ToggleInGameMenu()
+	//Singleplayer "Continue" counts as finished loading and shouldn't be affected
+	print("CancelLevelLoading has Initialized")
+	wait 30
+	if (uiGlobal.isLoading == true)
```
formatting :)
```suggestion
	if ( uiGlobal.isLoading == true )
```

uniboi commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,20 @@ void function UICodeCallback_ToggleInGameMenu()
+	wait 30
+	if (uiGlobal.isLoading == true)
+	{
+		printt("CancelLevelLoading disconnected because loading took too long")
```
unnecessary log that will just pollute logs
```suggestion
```

uniboi replied:
formatting
```suggestion
		printt( "CancelLevelLoading disconnected because loading took too long" )
```

uniboi commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,20 @@ void function UICodeCallback_ToggleInGameMenu()
 
+void function CancelLevelLoading()
+{
+	//Singleplayer "Continue" counts as finished loading and shouldn't be affected
```
I'd make it a constant for easy change
```suggestion
	const int timeoutThreshold = 30; // how long this function will wait before disconnecting

	//Singleplayer "Continue" counts as finished loading and shouldn't be affected
```

uniboi commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,20 @@ void function UICodeCallback_ToggleInGameMenu()
+{
+	//Singleplayer "Continue" counts as finished loading and shouldn't be affected
+	print("CancelLevelLoading has Initialized")
+	wait 30
```
assuming you use a constant
```suggestion
	wait timeoutThreshold
```

uniboi commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,20 @@ void function UICodeCallback_ToggleInGameMenu()
+		printt("CancelLevelLoading disconnected because loading took too long")
+		UICodeCallback_LevelLoadingFinished( true )
+		Disconnect()
+		OpenErrorDialog( "Loading timed out (Over 30 seconds)" )
```
assuming you use a constant
```suggestion
		OpenErrorDialog( format( "Loading timed out (Over %i seconds)", timeoutThreshold ) )
```

uniboi reviewed (changes requested):
I think 30 seconds before a timeout is fine

EnderBoy9217 commented:
Added the comments and uni's suggestions, also added a loop to reduce the chance of a false positive from loading multiple times
Unless anyone knows a way to cancel the script with another script without the loop

uniboi commented:
is there a point in aborting the function early? I'd just keep the `wait threshold` because letting the function run is cheaper than checking every second if the client is still loading.
Maybe there's a signal we can use, that'd be more elegant

ASpoonPlaysGames commented:
> Maybe there's a signal we can use, that'd be more elegant

there are various signals that might be of use (all of which are called on `uiGlobal.signalDummy`)

`LevelShutdown`
`LevelFinishedLoading`


i think `LevelShutdown` is best for this tbh

uniboi commented:
`LevelShutdown` triggers right before the map unloads iirc.
I'd make the function something like this

```
void function TimeoutHandler()
{
	const int timeoutThreshold = 30 // how long this function will wait before disconnecting
	EndSignal( uiGlobal.signalDummy, "LevelFinishedLoading" ) // cancel the thread if the level has finished loading

	wait timeoutThreshold;

	if ( uiGlobal.isLoading == true ) // Disconnect if the client still hasn't connected
	{
		printt( "TimeoutHandler disconnected because loading took too long" )
		UICodeCallback_LevelLoadingFinished( true ) // Removes loading screen
		Disconnect() // Loads main menu
		OpenErrorDialog( format( "Loading timed out (Over %i seconds)", timeoutThreshold ) ) // Opens error dialog to notify user
	}
}
```

EnderBoy9217 commented:
couldnt we get rid of the "if" statement if that signal ends the script
As that signal is sent right after isLoading becomes false

EnderBoy9217 commented:
endsignal causes the script to not work @uniboi , it will print that it is attempting to send the player back to lobby but it won't

Is there any other known way to end the script, or should we go back to the for loop

uniboi commented:
Try changing the function to this:

```
void function TimeoutHandler()
{
	const int timeoutThreshold = 30 // how long this function will wait before disconnecting
	EndSignal( uiGlobal.signalDummy, "LevelFinishedLoading" ) // cancel the thread if the level has finished loading

	OnThreadEnd(
		void function()
		{
				if ( uiGlobal.isLoading == true ) // Disconnect if the client still hasn't connected
				{
					printt( "TimeoutHandler disconnected because loading took too long" )
					UICodeCallback_LevelLoadingFinished( true ) // Removes loading screen
					Disconnect() // Loads main menu
					OpenErrorDialog( format( "Loading timed out (Over %i seconds)", timeoutThreshold ) ) // Opens error dialog to notify user
				}
		}
	)

	wait timeoutThreshold;
}
```

EnderBoy9217 commented:
I have no idea what you did there
But it didn't load the main menu ui, only the dialog and .bik file

EnderBoy9217 commented:
The current code still works but likely isn't the most optimal, if anyone has an alternative that doesn't use EndSignal (which seems to be causing issues) let me know

uniboi commented:
I'll take a look tomorrow

GeckoEidechse commented:
> Added the comments and uni's suggestions, also added a loop to reduce the chance of a false positive from loading multiple times.

Make sure to click the `Resolve` button on the conversations where it's applicable so we know that part has been addressed  ^^

F1F7Y commented:
> I'll take a look tomorrow

@uniboi bump

EnderBoy9217 commented:
Figured I should work on this again
@uniboi the problem was the `LevelFinishedLoading` had to be called before the disconnect, ending the thread so it never actually got to the disconnect


Changed it to `LevelShutdown` ( as @ASpoonPlaysGames recommended originally) and tested extensively, works (and now prevents the false kick if you happen to join and leave servers at a specific time)

Ready for review

ASpoonPlaysGames commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,24 @@ void function UICodeCallback_ToggleInGameMenu()
+{
+	//Singleplayer "Continue" counts as finished loading and shouldn't be affected
+	printt( "TimeoutHandler has Initialized" )
+	const int timeoutThreshold = 30 // how long this function will wait before disconnecting
```
Maybe this could be a ConVar?

EnderBoy9217 commented:
Added ConVar and per Spoon's request added it to modsettings

This also adds a Northstar.Client section to modsettings

### This PR now fixes an issue where modsettings cannot take an "int" ConVar

uniboi commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,23 @@ void function UICodeCallback_ToggleInGameMenu()
+void function TimeoutHandler()
+{
+	//Singleplayer "Continue" counts as finished loading and shouldn't be affected
+	printt( "TimeoutHandler has Initialized" )
```
```suggestion
```
useless log

ASpoonPlaysGames commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,23 @@ void function UICodeCallback_ToggleInGameMenu()
+	printt( "TimeoutHandler has Initialized" )
+	EndSignal( uiGlobal.signalDummy, "LevelShutdown" ) // Cancels thread if map is unloaded
+
+	wait GetConVarFloat( "ns_max_load_time" )
```
You use `GetConVarInt` elsewhere, is the ConVar an int or a float?

uniboi replied:
```suggestion
	float maxLoadTime = GetConVarFloat( "ns_max_load_time" )
	wait maxLoadTime
```

ASpoonPlaysGames commented on Northstar.Client/mod/scripts/vscripts/ui/menu_mod_settings.nut:
```diff
@@ -82,6 +82,9 @@ struct {
 void function AddModSettingsMenu()
 {
 	AddMenu( "ModSettings", $"resource/ui/menus/mod_settings.menu", InitModMenu )
+	AddModTitle( "^656dce00Northstar.Client" )
```
```suggestion

	AddModTitle( "^656dce00Northstar.Client" )
```
for clarity

uniboi commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,23 @@ void function UICodeCallback_ToggleInGameMenu()
+	wait GetConVarFloat( "ns_max_load_time" )
+	if ( uiGlobal.isLoading == true ) //Disconnects client after 30 seconds of loading
+	{
+		printt( "TimeoutHandler disconnected because loading took too long" )
```
```suggestion
		printt( format( "disconnecting after loading for %.1f seconds", maxLoadTime ) )
```

EnderBoy9217 replied:
LMK if you like the new print statement I put

ASpoonPlaysGames commented on Northstar.Client/mod/scripts/vscripts/ui/menu_mod_settings.nut:
```diff
@@ -343,6 +346,7 @@ void function UpdateListSliderHeight()
 
 void function UpdateList()
 {
+
```
```suggestion
```
Accidental change

uniboi commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,23 @@ void function UICodeCallback_ToggleInGameMenu()
+		printt( "TimeoutHandler disconnected because loading took too long" )
+		UICodeCallback_LevelLoadingFinished( true ) //Removes loading screen
+		Disconnect() // Loads menu
+		OpenErrorDialog( format( "Loading timed out (Over %i seconds)", GetConVarInt( "ns_max_load_time" ) ) ) //Opens error dialog to notify user
```
```suggestion
		OpenErrorDialog( format( "Loading timed out after %.1f seconds.\nYou can increase the limit in the Mod Settings on the Main Menu" ) ) ) // Opens error dialog to notify user
```
This should be localized though

EnderBoy9217 replied:
I am uploading that now, worded slightly differently

uniboi commented on Northstar.Client/mod/scripts/vscripts/ui/menu_mod_settings.nut:
```diff
@@ -82,6 +82,9 @@ struct {
 	AddMenu( "ModSettings", $"resource/ui/menus/mod_settings.menu", InitModMenu )
+	AddModTitle( "^656dce00Northstar.Client" )
+	AddModCategory( "#MISC" )
+	ModSettings_AddSetting( "ns_max_load_time", "#MAX_LOAD", "int" )
```
why not a float

EnderBoy9217 commented:
Changed it from an int to a float

ASpoonPlaysGames commented on Northstar.Client/mod.json:
```diff
@@ -1,10 +1,16 @@
 	"Name": "Northstar.Client",
 	"Description": "Various ui and client changes to fix bugs and add better support for mods",
-	"Version": "1.16.0",
+	"Version": "1.18.1",
```
Accidental change
```suggestion
	"Version": "1.16.0",
```

ASpoonPlaysGames commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,23 @@ void function UICodeCallback_ToggleInGameMenu()
+	wait maxLoadTime
+	if ( uiGlobal.isLoading == true ) //Disconnects client after 30 seconds of loading
+	{
+		printt( format( "TimeoutHandler disconnected because loading took too long (%i seconds)", maxLoadTime ) )
```
Since maxLoadTime is a float, does this work?
Might have to change it to this:
```suggestion
		printt( format( "TimeoutHandler disconnected because loading took too long (%f seconds)", maxLoadTime ) )
```

ASpoonPlaysGames commented on Northstar.Client/mod/scripts/vscripts/ui/menu_mod_settings.nut:
```diff
@@ -445,7 +449,7 @@ void function UpdateList()
 	file.updatingList = false
 
-	if ( file.conVarList.len() <= 0 )
+	if ( file.conVarList.len() <= 6 )
```
I assume this change was to make the "go download some mods" text show up now that Northstar itself is adding some settings right?

EnderBoy9217 replied:
@ASpoonPlaysGames yes

ASpoonPlaysGames commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,23 @@ void function UICodeCallback_ToggleInGameMenu()
+
+	float maxLoadTime = GetConVarFloat( "ns_max_load_time" )
+	wait maxLoadTime
+	if ( uiGlobal.isLoading == true ) //Disconnects client after 30 seconds of loading
```
```suggestion
	if ( uiGlobal.isLoading ) //Disconnects client after 30 seconds of loading
```
Comparing bool to true is unnecessary

ASpoonPlaysGames commented on Northstar.Client/mod/scripts/vscripts/ui/_menus.nut:
```diff
@@ -178,6 +178,23 @@ void function UICodeCallback_ToggleInGameMenu()
+		printt( format( "TimeoutHandler disconnected because loading took too long (%i seconds)", maxLoadTime ) )
+		UICodeCallback_LevelLoadingFinished( true ) //Removes loading screen
+		Disconnect() // Loads menu
+		OpenErrorDialog( format( Localize("#MAX_LOAD_ERROR"), maxLoadTime ) ) //Opens error dialog to notify user
```
```suggestion
		OpenErrorDialog( format( Localize( "#MAX_LOAD_ERROR" ), maxLoadTime ) ) //Opens error dialog to notify user
```
Formatting

ASpoonPlaysGames reviewed (changes requested):
Just a few more things from me

ASpoonPlaysGames reviewed (approved):
Code looks good :)

EnderBoy9217 commented:
I got an odd error while playing with this and I think its related
![image](https://github.com/R2Northstar/NorthstarMods/assets/122132914/1a698098-d362-42e6-80bf-c17154986c0d)
[nslog2023-08-27 12-09-30.txt](https://github.com/R2Northstar/NorthstarMods/files/12448379/nslog2023-08-27.12-09-30.txt)

ASpoonPlaysGames commented:
![image](https://github.com/R2Northstar/NorthstarMods/assets/66967891/cef8237d-20db-4e20-9d02-a9a38aaa195b)
indeed it is

EnderBoy9217 commented:
I believe its fixed with an "IsConnected()" check I just added, works fine in testing

EnderBoy9217 commented:
Also this fix works if the masterserver was to go down while someone was loading for some reason

F1F7Y reviewed (changes requested):
Blocking as:
1. Each fix should be a separate PR
2. The "fix" for infinite loading screen is really hacky and should be handled in native by checking the connection like valve source does.
